// @nteract/commutable/lib/structures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const primitives_1 = require("./primitives");
const notebook_1 = require("./notebook");
const cells_1 = require("./cells");
// The cell creators here are a bit duplicative
exports.createCodeCell = cells_1.makeCodeCell;
exports.createMarkdownCell = cells_1.makeMarkdownCell;
exports.emptyCodeCell = exports.createCodeCell();
exports.emptyMarkdownCell = exports.createMarkdownCell();
// These are all kind of duplicative now that we're on records.
// Since we export these though, they're left for
// backwards compatiblity
exports.defaultNotebook = notebook_1.makeNotebookRecord();
exports.createNotebook = notebook_1.makeNotebookRecord;
exports.emptyNotebook = notebook_1.makeNotebookRecord();
/**
 * A function that appends a new cell to a CellStructure object.
 *
 * @param cellStructure The cellOrder and cellMap of the current notebook
 * @param immutableCell The cell that will be inserted into the cellStructure
 * @param id The id of the new cell, defaults to a new UUID
 *
 * @returns Cell structure with the new cell appended at the end
 */
exports.appendCell = (cellStructure, immutableCell, id = primitives_1.createCellId()) => ({
    cellOrder: cellStructure.cellOrder.push(id),
    cellMap: cellStructure.cellMap.set(id, immutableCell)
});
/**
 * A function that appends a cell to an immutable notebook.
 *
 * @param immnb An immutable data structure representing the notebook that will be modified
 * @param immCell The new cell that will be inserted into the notebook
 *
 * @returns The modified notebook
 */
exports.appendCellToNotebook = (immnb, immCell) => immnb.withMutations(nb => {
    const cellStructure = {
        cellOrder: nb.get("cellOrder"),
        cellMap: nb.get("cellMap")
    };
    const { cellOrder, cellMap } = exports.appendCell(cellStructure, immCell);
    return nb.set("cellOrder", cellOrder).set("cellMap", cellMap);
});
/**
 * Inserts a cell with cellID at a given index within the notebook.
 *
 * @param notebook The notebook the cell will be inserted into.
 * @param cell The cell that will be inserted
 * @param cellID The ID of the cell.
 * @param index The position we would like to insert the cell at
 *
 * @returns The modified notebook.
 */
exports.insertCellAt = (notebook, cell, cellId, index) => notebook.withMutations(nb => nb
    .setIn(["cellMap", cellId], cell)
    .set("cellOrder", nb.get("cellOrder").insert(index, cellId)));
/**
 * Inserts a new cell with cellID before an existing cell with priorCellID
 * in the notebook.
 *
 * @param notebook The notebook the cell will be inserted into.
 * @param cell The cell that will be inserted
 * @param cellID The ID of the cell.
 * @param priorCellID The ID of the existing cell.
 */
exports.insertCellAfter = (notebook, cell, cellId, priorCellId) => exports.insertCellAt(notebook, cell, cellId, notebook.get("cellOrder").indexOf(priorCellId) + 1);
/**
 * Delete a cell with CellID at a given location. Note that this function
 * is deprecated in favor of `deleteCell`.
 *
 * @param notebook The notebook containing the cell.
 * @param cellID The ID of the cell that will be deleted.
 *
 * @returns The modified notebook
 *
 * @deprecated use `deleteCell()` instead
 */
exports.removeCell = (notebook, cellId) => {
    console.log("Deprecation Warning: removeCell() is being deprecated. Please use deleteCell() instead");
    return exports.deleteCell(notebook, cellId);
};
/**
 * Delete a cell with CellID at a given location.
 *
 * @param notebook The notebook containing the cell.
 * @param cellID The ID of the cell that will be deleted.
 *
 * @returns The modified notebook
 */
exports.deleteCell = (notebook, cellId) => notebook
    .removeIn(["cellMap", cellId])
    .update("cellOrder", cellOrder => cellOrder.filterNot(id => id === cellId));
/**
 * A new notebook with a single empty code cell. This function is useful
 * if you are looking to initialize a fresh, new notebook.
 */
exports.monocellNotebook = exports.appendCellToNotebook(exports.emptyNotebook, exports.emptyCodeCell);
