// @nteract/types/lib/entities/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @module types
 */
const Immutable = __importStar(require("immutable"));
const contents_1 = require("./contents");
const hosts_1 = require("./hosts");
const kernels_1 = require("./kernels");
const kernelspecs_1 = require("./kernelspecs");
const modals_1 = require("./modals");
__export(require("./contents"));
__export(require("./hosts"));
__export(require("./kernels"));
__export(require("./kernel-info"));
__export(require("./kernelspecs"));
__export(require("./modals"));
exports.makeEntitiesRecord = Immutable.Record({
    contents: contents_1.makeContentsRecord(),
    hosts: hosts_1.makeHostsRecord(),
    kernels: kernels_1.makeKernelsRecord(),
    kernelspecs: kernelspecs_1.makeKernelspecsRecord(),
    modals: modals_1.makeModalsRecord()
});
