// @nteract/types/lib/entities/kernels.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const Immutable = __importStar(require("immutable"));
const rxjs_1 = require("rxjs");
exports.makeKernelNotStartedRecord = Immutable.Record({
    kernelSpecName: null,
    status: null,
    type: "unknown",
    cwd: ".",
    channels: new rxjs_1.Subject(),
    info: null
});
exports.makeLocalKernelRecord = Immutable.Record({
    type: "zeromq",
    cwd: ".",
    info: null,
    kernelSpecName: null,
    hostRef: null,
    lastActivity: null,
    channels: new rxjs_1.Subject(),
    status: null,
    spawn: null,
    connectionFile: null
});
exports.makeRemoteKernelRecord = Immutable.Record({
    type: "websocket",
    info: null,
    cwd: ".",
    id: null,
    kernelSpecName: null,
    hostRef: null,
    lastActivity: null,
    channels: new rxjs_1.Subject(),
    sessionId: null,
    status: null
});
exports.makeKernelsRecord = Immutable.Record({
    byRef: Immutable.Map()
});
