// @nteract/types/lib/entities/kernel-info.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @module types
 */
const Immutable = __importStar(require("immutable"));
exports.makeHelpLinkRecord = Immutable.Record({
    text: "",
    url: "#"
});
// NOTE: This is not exported as the "makeXRecord" for KernelInfo so that we
//       don't have nullables and can rely on the proper types. The developer is
//       required to pass a fully instantiated KernelInfo into
//       makeKernelInfoRecord below.
const kernelInfoRecordMaker = Immutable.Record({
    protocolVersion: "0.0.0",
    implementation: "",
    implementationVersion: "",
    banner: "",
    helpLinks: Immutable.List(),
    languageName: "",
    languageVersion: "",
    mimetype: "",
    fileExtension: "",
    pygmentsLexer: "",
    codemirrorMode: "text/plain",
    nbconvertExporter: ""
});
function makeKernelInfoRecord(kernelInfo) {
    return kernelInfoRecordMaker(kernelInfo);
}
exports.makeKernelInfoRecord = makeKernelInfoRecord;
