// ansi-to-react/lib/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const anser_1 = require("anser");
const escape_carriage_1 = require("escape-carriage");
const LINK_REGEX = /(https?:\/\/(?:www\.|(?!www))[^\s\.]+\.[^\s]{2,}|www\.[^\s]+\.[^\s]{2,})/;
/**
 * Converts ANSI strings into JSON output.
 * @name ansiToJSON
 * @function
 * @param {String} input The input string.
 * @return {Array} The parsed input.
 */
function ansiToJSON(input) {
    input = escape_carriage_1.escapeCarriageReturn(input);
    return anser_1.ansiToJson(input, {
        json: true,
        remove_empty: true
    });
}
/**
 * Converts an Anser bundle into a React Node.
 * @param linkify whether links should be converting into clickable anchor tags.
 * @param bundle Anser output.
 */
function convertBundleIntoReact(linkify, bundle, key) {
    const style = {};
    if (bundle.bg) {
        style.backgroundColor = `rgb(${bundle.bg})`;
    }
    if (bundle.fg) {
        style.color = `rgb(${bundle.fg})`;
    }
    if (!linkify) {
        return React.createElement("span", { style, key }, bundle.content);
    }
    const words = bundle.content.split(" ").reduce((words, word, index) => {
        // If this isn't the first word, re-add the space removed from split.
        if (index !== 0) {
            words.push(" ");
        }
        // If  this isn't a link, just return the word as-is.
        if (!LINK_REGEX.test(word)) {
            words.push(word);
            return words;
        }
        words.push(React.createElement("a", {
            key: index,
            href: word,
            target: "_blank"
        }, `${word}`));
        return words;
    }, []);
    return React.createElement("span", { style, key }, words);
}
function Ansi(props) {
    return React.createElement("code", { className: props.className }, ansiToJSON(props.children).map(convertBundleIntoReact.bind(null, props.linkify)));
}
exports.default = Ansi;
