// @nteract/records/lib/outputs/display-data.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const immer_1 = __importDefault(require("immer"));
const common = __importStar(require("../common"));
exports.DISPLAYDATA = "display_data";
exports.outputType = exports.DISPLAYDATA;
function displayData(displayDataOutput) {
    const defaultDisplayData = {
        outputType: exports.DISPLAYDATA,
        data: {},
        metadata: {}
    };
    return immer_1.default(defaultDisplayData, draft => {
        return Object.assign(draft, displayDataOutput);
    });
}
exports.displayData = displayData;
displayData.type = "display_data";
displayData.fromNbformat = function fromNbformat(s) {
    return displayData({
        outputType: s.output_type,
        data: common.createImmutableMimeBundle(s.data),
        metadata: s.metadata
    });
};
displayData.fromJupyterMessage = function displayDataRecordFromMessage(msg) {
    return displayData({
        outputType: exports.DISPLAYDATA,
        // The data field in a display data output type on the message spec is the same as we need
        // We could do additional checking here though
        data: msg.content.data,
        metadata: msg.content.metadata
    });
};
