// @nteract/records/lib/outputs/execute-result.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const immer_1 = __importDefault(require("immer"));
const common = __importStar(require("../common"));
exports.EXECUTE_RESULT = "execute_result";
function executeResult(executeResultOutput) {
    const defaultExecuteResult = {
        outputType: exports.EXECUTE_RESULT,
        executionCount: undefined,
        data: {},
        metadata: {}
    };
    return immer_1.default(defaultExecuteResult, draft => {
        return Object.assign(draft, executeResultOutput);
    });
}
exports.executeResult = executeResult;
executeResult.type = exports.EXECUTE_RESULT;
executeResult.fromNbformat = function fromNbformat(s) {
    return executeResult({
        executionCount: s.execution_count,
        data: common.createImmutableMimeBundle(s.data),
        metadata: s.metadata
    });
};
executeResult.fromJupyterMessage = function fromJupyterMessage(msg) {
    return executeResult({
        executionCount: msg.content.execution_count,
        data: msg.content.data,
        metadata: msg.content.metadata
    });
};
