// @nteract/records/lib/outputs/stream.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const common = __importStar(require("../common"));
exports.STDOUT = "stdout";
exports.STDERR = "stderr";
exports.STREAM = "stream";
function streamOutput(s) {
    return Object.freeze(Object.assign({}, { outputType: exports.STREAM, name: exports.STDOUT, text: "" }, s));
}
exports.streamOutput = streamOutput;
streamOutput.type = exports.STREAM;
streamOutput.fromNbformat = function fromNbformat(s) {
    return streamOutput({
        name: s.name,
        text: common.demultiline(s.text)
    });
};
streamOutput.fromJupyterMessage = function fromJupyterMessage(msg) {
    return streamOutput({
        name: msg.content.name,
        text: msg.content.text
    });
};
