// gatsby/dist/state-machines/query-running/index.js
"use strict";

exports.__esModule = true;
exports.queryRunningMachine = exports.queryStates = void 0;

var _xstate = require("xstate");

var _services = require("./services");

var _actions = require("./actions");

/**
 * This is a child state machine, spawned to perform the query running
 */
const queryStates = {
  initial: `extractingQueries`,
  id: `queryRunningMachine`,
  on: {
    SOURCE_FILE_CHANGED: {
      actions: `markSourceFilesDirty`
    }
  },
  context: {},
  states: {
    extractingQueries: {
      id: `extracting-queries`,
      invoke: {
        id: `extracting-queries`,
        src: `extractQueries`,
        onDone: {
          target: `writingRequires`,
          actions: `markSourceFilesClean`
        }
      }
    },
    writingRequires: {
      invoke: {
        src: `writeOutRequires`,
        id: `writing-requires`,
        onDone: {
          target: `calculatingDirtyQueries`
        }
      }
    },
    calculatingDirtyQueries: {
      invoke: {
        id: `calculating-dirty-queries`,
        src: `calculateDirtyQueries`,
        onDone: {
          target: `runningStaticQueries`,
          actions: `assignDirtyQueries`
        }
      }
    },
    runningStaticQueries: {
      invoke: {
        src: `runStaticQueries`,
        id: `running-static-queries`,
        onDone: {
          target: `runningPageQueries`
        }
      }
    },
    runningPageQueries: {
      invoke: {
        src: `runPageQueries`,
        id: `running-page-queries`,
        onDone: {
          target: `waitingForJobs`,
          actions: `flushPageData`
        }
      }
    },
    // This waits for the jobs API to finish
    waitingForJobs: {
      // If files are dirty go and extract again
      always: {
        cond: ctx => !!ctx.filesDirty,
        target: `extractingQueries`
      },
      invoke: {
        src: `waitUntilAllJobsComplete`,
        id: `waiting-for-jobs`,
        onDone: {
          target: `done`
        }
      }
    },
    done: {
      type: `final`
    }
  }
};
exports.queryStates = queryStates;
const queryRunningMachine = (0, _xstate.Machine)(queryStates, {
  actions: _actions.queryActions,
  services: _services.queryRunningServices
});
exports.queryRunningMachine = queryRunningMachine;
