// @nteract/presentational-components/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolbarIcons = exports.ToolbarItem = exports.Toolbar = exports.SidebarItem = exports.Sidebar = exports.Error = exports.Cells = exports.Cell = exports.Source = exports.PromptBuffer = exports.Prompt = exports.Pagers = exports.Outputs = exports.Input = exports.DarkTheme = exports.LightTheme = exports.ToggleSwitch = exports.Icons = exports.CellMenuSection = exports.CellMenuItem = exports.CellMenu = exports.CircularButton = exports.CommandPalette = exports.CommandButtonRow = void 0;
const cell_1 = require("./components/cell");
Object.defineProperty(exports, "Cell", { enumerable: true, get: function () { return cell_1.Cell; } });
const cells_1 = require("./components/cells");
Object.defineProperty(exports, "Cells", { enumerable: true, get: function () { return cells_1.Cells; } });
const CircularButton_1 = require("./components/CircularButton");
Object.defineProperty(exports, "CircularButton", { enumerable: true, get: function () { return CircularButton_1.CircularButton; } });
const error_1 = __importDefault(require("./components/error"));
exports.Error = error_1.default;
const input_1 = require("./components/input");
Object.defineProperty(exports, "Input", { enumerable: true, get: function () { return input_1.Input; } });
const outputs_1 = require("./components/outputs");
Object.defineProperty(exports, "Outputs", { enumerable: true, get: function () { return outputs_1.Outputs; } });
const pagers_1 = require("./components/pagers");
Object.defineProperty(exports, "Pagers", { enumerable: true, get: function () { return pagers_1.Pagers; } });
const prompt_1 = require("./components/prompt");
Object.defineProperty(exports, "Prompt", { enumerable: true, get: function () { return prompt_1.Prompt; } });
Object.defineProperty(exports, "PromptBuffer", { enumerable: true, get: function () { return prompt_1.PromptBuffer; } });
const sidebar_1 = require("./components/sidebar");
Object.defineProperty(exports, "Sidebar", { enumerable: true, get: function () { return sidebar_1.Sidebar; } });
Object.defineProperty(exports, "SidebarItem", { enumerable: true, get: function () { return sidebar_1.SidebarItem; } });
const source_1 = require("./components/source");
Object.defineProperty(exports, "Source", { enumerable: true, get: function () { return source_1.Source; } });
const CellMenu_1 = require("./components/CellMenu");
Object.defineProperty(exports, "CellMenu", { enumerable: true, get: function () { return CellMenu_1.CellMenu; } });
Object.defineProperty(exports, "CellMenuItem", { enumerable: true, get: function () { return CellMenu_1.CellMenuItem; } });
Object.defineProperty(exports, "CellMenuSection", { enumerable: true, get: function () { return CellMenu_1.CellMenuSection; } });
const ToggleSwitch_1 = require("./components/ToggleSwitch/ToggleSwitch");
Object.defineProperty(exports, "ToggleSwitch", { enumerable: true, get: function () { return ToggleSwitch_1.ToggleSwitch; } });
const styles_1 = require("./styles");
Object.defineProperty(exports, "DarkTheme", { enumerable: true, get: function () { return styles_1.DarkTheme; } });
Object.defineProperty(exports, "LightTheme", { enumerable: true, get: function () { return styles_1.LightTheme; } });
const command_palette_1 = require("./components/command-palette");
Object.defineProperty(exports, "CommandButtonRow", { enumerable: true, get: function () { return command_palette_1.CommandButtonRow; } });
Object.defineProperty(exports, "CommandPalette", { enumerable: true, get: function () { return command_palette_1.CommandPalette; } });
const Icons = __importStar(require("./components/icons"));
exports.Icons = Icons;
__exportStar(require("./styles"), exports);
const toolbar_1 = require("./components/toolbar");
Object.defineProperty(exports, "Toolbar", { enumerable: true, get: function () { return toolbar_1.Toolbar; } });
Object.defineProperty(exports, "ToolbarItem", { enumerable: true, get: function () { return toolbar_1.ToolbarItem; } });
const ToolbarIcons = {
    Clear: toolbar_1.Clear,
    Commands: toolbar_1.Commands,
    Expand: toolbar_1.Expand,
    Collapse: toolbar_1.Collapse,
    ExpandSidebar: toolbar_1.ExpandSidebar,
    Group: toolbar_1.Group,
    Pause: toolbar_1.Pause,
    Play: toolbar_1.Play,
    Plus: toolbar_1.Plus,
};
exports.ToolbarIcons = ToolbarIcons;
