// @nteract/mathjax/lib/provider.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const load_script_1 = __importDefault(require("./load-script"));
const context_1 = __importDefault(require("./context"));
/**
 * MathJax Provider
 */
class Provider extends React.Component {
    constructor(props) {
        super(props);
        this.onLoad = () => {
            if (typeof MathJax === "undefined" || !MathJax || !MathJax.Hub) {
                this.props.onError(new Error("MathJax not really loaded even though onLoad called"));
                return;
            }
            const options = this.props.options;
            if (options != null && Object.keys(options).length > 0) {
                MathJax.Hub.Config(options);
            }
            MathJax.Hub.Register.StartupHook("End", () => {
                if (typeof MathJax === "undefined" || !MathJax) {
                    this.props.onError(new Error("MathJax became undefined in the middle of processing"));
                    return;
                }
                MathJax.Hub.processSectionDelay = this.props.delay;
                if (this.props.didFinishTypeset) {
                    this.props.didFinishTypeset();
                }
            });
            MathJax.Hub.Register.MessageHook("Math Processing Error", (message) => {
                if (this.props.onError) {
                    this.props.onError(new Error(message));
                }
            });
            if (this.props.onLoad) {
                this.props.onLoad();
            }
            this.setState({
                MathJax
            });
        };
        this.state = {
            MathJax: undefined,
            input: this.props.input,
            hasProviderAbove: true
        };
    }
    static getDerivedStateFromProps(props, state) {
        if (state.input !== props.input) {
            return Object.assign({}, state, { input: props.input });
        }
        return null;
    }
    componentDidMount() {
        const src = this.props.src;
        if (src == null) {
            return this.onLoad();
        }
        if (typeof MathJax === "undefined" || !MathJax || !MathJax.Hub) {
            return load_script_1.default(src, this.onLoad);
        }
        this.onLoad();
    }
    render() {
        return (React.createElement(context_1.default.Provider, { value: this.state }, this.props.children));
    }
}
Provider.defaultProps = {
    src: "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML",
    input: "tex",
    didFinishTypeset: null,
    delay: 0,
    options: {},
    loading: null,
    noGate: false,
    onLoad: null,
    onError: (err) => {
        console.error(err);
    }
};
exports.default = Provider;
