// @nteract/types/lib/entities/contents/notebook.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @module types
 */
const Immutable = __importStar(require("immutable"));
const commutable_1 = require("@nteract/commutable");
exports.makeDocumentRecord = Immutable.Record({
    type: "notebook",
    notebook: commutable_1.emptyNotebook,
    savedNotebook: commutable_1.emptyNotebook,
    transient: Immutable.Map({
        keyPathsForDisplays: Immutable.Map()
    }),
    cellPagers: Immutable.Map(),
    editorFocused: null,
    cellFocused: null,
    copied: null,
    kernelRef: null
});
exports.makeNotebookContentRecord = Immutable.Record({
    mimetype: null,
    created: null,
    format: "json",
    lastSaved: null,
    model: exports.makeDocumentRecord(),
    filepath: "",
    type: "notebook",
    writable: true
});
