// @nteract/records/lib/cells/code-cell.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const immer_1 = __importDefault(require("immer"));
const outputs_1 = require("../outputs");
exports.CODECELL = "code";
exports.makeCodeCellRecord = (codeCell) => {
    const defaultCodeCell = {
        cellType: "code",
        metadata: {},
        executionCount: undefined,
        source: "",
        outputs: []
    };
    return immer_1.default(defaultCodeCell, draft => Object.assign(draft, codeCell));
};
function codeCellRecordFromNbformat(codeCell) {
    return exports.makeCodeCellRecord({
        cellType: codeCell.cell_type,
        metadata: codeCell.metadata,
        executionCount: codeCell.execution_count,
        source: codeCell.source,
        outputs: codeCell.outputs.map(outputs_1.outputFromNbformat)
    });
}
exports.codeCellRecordFromNbformat = codeCellRecordFromNbformat;
