// @nteract/records/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./outputs"));
__export(require("./cells"));
__export(require("./common"));
const outputs = __importStar(require("./outputs"));
// Deprecation Warning: removeCell is being deprecated. Please use deleteCell instead
const { emptyNotebook, emptyCodeCell, emptyMarkdownCell, appendCell, monocellNotebook, createCodeCell, appendCellToNotebook, insertCellAt, insertCellAfter, removeCell, deleteCell } = require("./structures");
exports.emptyNotebook = emptyNotebook;
exports.emptyCodeCell = emptyCodeCell;
exports.emptyMarkdownCell = emptyMarkdownCell;
exports.appendCell = appendCell;
exports.monocellNotebook = monocellNotebook;
exports.createCodeCell = createCodeCell;
exports.appendCellToNotebook = appendCellToNotebook;
exports.insertCellAt = insertCellAt;
exports.insertCellAfter = insertCellAfter;
exports.removeCell = removeCell;
exports.deleteCell = deleteCell;
function freezeReviver(k, v) {
    return Object.freeze(v);
}
function parseNotebook(notebookString) {
    return JSON.parse(notebookString, freezeReviver);
}
exports.parseNotebook = parseNotebook;
// Expected usage is stringifyNotebook(toJS(immutableNotebook))
function stringifyNotebook(notebook) {
    return JSON.stringify(notebook, null, 2);
}
exports.stringifyNotebook = stringifyNotebook;
const createImmutableOutput = outputs.outputFromNbformat;
exports.createImmutableOutput = createImmutableOutput;
